File: src/lib/worker/embed.ts
/**
 * Thin wrapper around OpenAI's text-embedding-3-small.
 *
 * Why raw fetch (not @ai-sdk/openai or the openai SDK):
 *   - One endpoint, well-specified, ~30 lines. Avoids a new dep.
 *   - Env-gated: returns null cleanly when OPENAI_API_KEY is unset, so the
 *     rest of the catalog still works (lexical-only retrieval) until a key
 *     is provided.
 *   - Batches up to BATCH inputs per call. text-embedding-3-small accepts
 *     arrays — one network roundtrip per 100 inputs is the cost-optimal mode.
 *
 * Pricing (as of 2026-01): $0.02 / 1M tokens. 10K tables × ~500 tokens ≈ $0.10
 * cold backfill. Per-chat-turn query embed (~50 tokens) is ~$0.000001 — free.
 */
import { env } from "../shared/env";
import { log } from "../shared/log";

const MODEL = "text-embedding-3-small";
const DIM = 1536;
const BATCH = 96;            // OpenAI accepts up to 2048 inputs but ~100 is the latency sweet spot
const ENDPOINT = "https://api.openai.com/v1/embeddings";

export function embeddingsEnabled(): boolean {
  return Boolean(env().OPENAI_API_KEY);
}

export const EMBEDDING_DIM = DIM;

/**
 * Embed one string. Returns null if no key is configured.
 * Use embedMany for >1 input — batching is cheaper.
 */
export async function embedOne(text: string): Promise<number[] | null> {
  const arr = await embedMany([text]);
  return arr?.[0] ?? null;
}

/**
 * Embed an array of strings. Returns null if no key is configured.
 * Output is parallel-indexed to the input array.
 */
export async function embedMany(texts: string[]): Promise<number[][] | null> {
  const key = env().OPENAI_API_KEY;
  if (!key) return null;
  if (texts.length === 0) return [];

  const out: number[][] = new Array(texts.length);
  for (let i = 0; i < texts.length; i += BATCH) {
    const slice = texts.slice(i, i + BATCH);
    const vecs = await callBatch(key, slice);
    for (let j = 0; j < vecs.length; j++) {
      out[i + j] = vecs[j]!;
    }
  }
  return out;
}

async function callBatch(key: string, inputs: string[]): Promise<number[][]> {
  const t0 = Date.now();
  const res = await fetch(ENDPOINT, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      Authorization: `Bearer ${key}`,
    },
    body: JSON.stringify({ model: MODEL, input: inputs }),
  });
  if (!res.ok) {
    const body = await res.text().catch(() => "");
    throw new Error(`OpenAI embeddings ${res.status}: ${body.slice(0, 200)}`);
  }
  const json = (await res.json()) as {
    data: { index: number; embedding: number[] }[];
    usage?: { prompt_tokens?: number; total_tokens?: number };
  };
  // OpenAI guarantees the response array order, but we sort by index to be safe
  // against intermediate proxies that may reorder.
  const sorted = [...json.data].sort((a, b) => a.index - b.index);
  log.info("embed.batch", {
    n: inputs.length,
    tokens: json.usage?.total_tokens ?? 0,
    ms: Date.now() - t0,
  });
  return sorted.map((d) => d.embedding);
}

/**
 * Format a pgvector literal: `[0.1,0.2,...]`. pg's parameter binding stringifies
 * arrays with curly braces by default, which pgvector rejects.
 */
export function toPgVectorLiteral(vec: number[]): string {
  return `[${vec.join(",")}]`;
}

File: backend/app/workers/embed.py
"""OpenAI embeddings wrapper.

Port of src/lib/worker/embed.ts. Returns None cleanly when OPENAI_API_KEY is
unset, so the rest of the catalog still works (lexical-only retrieval) until
a key is provided.
"""
from __future__ import annotations

import logging

import httpx

from ..config import get_settings

log = logging.getLogger(__name__)

MODEL = "text-embedding-3-small"
EMBEDDING_DIM = 1536
_BATCH = 96
_ENDPOINT = "https://api.openai.com/v1/embeddings"


def embeddings_enabled() -> bool:
    return bool(get_settings().openai_api_key)


async def embed_one(text: str) -> list[float] | None:
    vecs = await embed_many([text])
    return vecs[0] if vecs else None


async def embed_many(texts: list[str]) -> list[list[float]] | None:
    key = get_settings().openai_api_key
    if not key:
        return None
    if not texts:
        return []

    out: list[list[float]] = []
    async with httpx.AsyncClient(timeout=30) as client:
        for i in range(0, len(texts), _BATCH):
            chunk = texts[i : i + _BATCH]
            vecs = await _call_batch(client, key, chunk)
            out.extend(vecs)
    return out


async def _call_batch(client: httpx.AsyncClient, key: str, inputs: list[str]) -> list[list[float]]:
    res = await client.post(
        _ENDPOINT,
        headers={"Authorization": f"Bearer {key}", "Content-Type": "application/json"},
        json={"model": MODEL, "input": inputs},
    )
    if res.status_code >= 400:
        body = res.text[:200]
        raise RuntimeError(f"OpenAI embeddings {res.status_code}: {body}")
    payload = res.json()
    rows = sorted(payload["data"], key=lambda d: d["index"])
    log.info("embed.batch n=%s tokens=%s", len(inputs),
             (payload.get("usage") or {}).get("total_tokens", 0))
    return [row["embedding"] for row in rows]


def to_pgvector_literal(vec: list[float]) -> str:
    return "[" + ",".join(str(v) for v in vec) + "]"
